### PR TITLE
SAMZA-2560: Generate SamzaSQLRelRecord using avro schema of input event instead of cached-schema.

### PR DESCRIPTION
**Pull request by shanthoosh**
*Symptom:* Input messages consumed by the Samza SQL queries were incorrectly serialized.  

*Cause:* Samza SQL engine uses `AvroRelConverter` to convert the input messages from Avro-type to `SamzaSqlRelRecord` before they're processed by the high-level operators in samza. Currently `AvroRelConverter` incorrectly converts the nested non-nullable unions in the input messages. If the schema of input event contains nested non-nullable union, then `AvroRelConverter` always produces null values for corresponding fields in resultant `SamzaSQLRelRecord`.  This problem happens since cached schema of input source is used to convert the input event to `SamzaSqlRelRecord`.

*Changes:* Use the input event schema to convert the event to SamzaSqlRelRecord rather than using the cached-schema.

*API changes:* None

*Upgrade Instructions:* None

*Usage Instructions:* None

### CODE REVIEW

**Comment by shanthoosh** [comment]
@atoomula Can you please take a look at these changes.

File: samza-sql/src/main/java/org/apache/samza/sql/interfaces/SqlIOConfig.java [hunk 1]
```diff
@@ -81,11 +81,13 @@ public SqlIOConfig(String systemName, String streamName, List<String> sourcePart
     this.streamId = String.format("%s-%s", systemName, streamName);
 
     samzaRelConverterName = streamConfigs.get(CFG_SAMZA_REL_CONVERTER);
-    Validate.notEmpty(samzaRelConverterName, String.format("System %s is not supported. Please check if the system name is provided correctly.", systemName));
```
Comment (atoomula) [inline samza-sql/src/main/java/org/apache/samza/sql/interfaces/SqlIOConfig.java:84]: Can you revert the change for validation here and below ?

Comment (shanthoosh) [reply samza-sql/src/main/java/org/apache/samza/sql/interfaces/SqlIOConfig.java:84]: Done.

File: samza-sql/src/main/java/org/apache/samza/sql/avro/AvroRelConverter.java [hunk 2]
```diff
@@ -125,11 +125,18 @@ public static void fetchFieldNamesAndValuesFromIndexedRecord(IndexedRecord recor
         .collect(Collectors.toList()));
   }
 
-  private static SamzaSqlRelRecord convertToRelRecord(IndexedRecord avroRecord, Schema schema) {
+  private static SamzaSqlRelRecord convertToRelRecord(IndexedRecord avroRecord) {
     List<Object> fieldValues = new ArrayList<>();
     List<String> fieldNames = new ArrayList<>();
     if (avroRecord != null) {
-      fetchFieldNamesAndValuesFromIndexedRecord(avroRecord, fieldNames, fieldValues, schema);
+      fieldNames.addAll(
```
Comment (b-slim) [inline samza-sql/src/main/java/org/apache/samza/sql/avro/AvroRelConverter.java:132]: Can we add a simple unit test or at least a comment to avoid that this happens again in the future ?

Comment (shanthoosh) [reply samza-sql/src/main/java/org/apache/samza/sql/avro/AvroRelConverter.java:132]: Done. Added unit tests for Avro record conversion with non-nullable union type.

File: samza-sql/src/test/java/org/apache/samza/sql/avro/schemas/ComplexUnion.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
+/**
+ * Autogenerated by Avro
+ * 
+ * DO NOT EDIT DIRECTLY
+ */
+package org.apache.samza.sql.avro.schemas;
```
Comment (b-slim) [inline samza-sql/src/test/java/org/apache/samza/sql/avro/schemas/ComplexUnion.java]: I am wondering if this can be auto generated by the Build/Testing framework at runtime ? using something like `avro-maven-plugin`  here is an example https://github.com/alexholmes/avro-maven.
FYI this is a suggestion comment and can be tackled as a followup if it is has a big scope.

Comment (shanthoosh) [reply samza-sql/src/test/java/org/apache/samza/sql/avro/schemas/ComplexUnion.java]: I agree that auto-generating the avro-code using gradle plugin at build-time would be better and i'll create a follow-up ticket for it.

File: samza-sql/src/test/java/org/apache/samza/sql/avro/TestAvroRelConversion.java [hunk 4]
```diff
@@ -231,6 +240,42 @@ public void testComplexRecordConversion() throws IOException {
     validateAvroSerializedData(serializedData, testStrValue);
   }
 
+  @Test
+  public void testComplexUnionConversionShouldWorkWithBothStringAndIntTypes() throws Exception {
```
Comment (atoomula) [inline samza-sql/src/test/java/org/apache/samza/sql/avro/TestAvroRelConversion.java:244]: Were you able to repro the issue with this unit test by having the older changes (with cached schema) ?
